man/man5/rlm_attr_filter.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_attr_filter 5 "27 June 2013" "" "FreeRADIUS Module"
.SH NAME
rlm_attr_filter \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_attr_filter\fP module exists for filtering certain
attributes and values in received ( or transmitted ) radius packets.
It gives the server a flexible framework to filter the attributes we
send to or receive from home servers or NASes.  This makes sense, for
example, in an out-sourced dialup situation to various policy
decisions, such as restricting a client to certain ranges of
Idle-Timeout or Session-Timeout.
.PP
Filter rules are normally defined and applied on a per-realm basis,
Filter rules can optionally be applied using another attribute, by
editing the \fIkey\fP configuration for this module.
.PP
In 2.0.1 and earlier versions, the "accounting" section filtered the
Accounting-Request, even though it was documented as filtering the
response.  This issue has been fixed in version 2.0.2 and later
versions.  The "preacct" section may now be used to filter
Accounting-Request packets.  The "accounting" section now filters
Accounting-Response packets.  Administrators using "attr_filter" in
the "accounting" section SHOULD move the reference to "attr_filter"
from "accounting" to "preacct".
.PP
The file that defines the attribute filtering rules follows a similar
syntax to the \fIusers\fP file.  There are a few differences however:
.PP
.DS
    There are no check-items allowed other than the name of the key.
.PP
    There can only be a single DEFAULT entry.
.PP
The rules for each entry are parsed to top to bottom, and an
attribute must pass *all* the rules which affect it in order to
make it past the filter.  Order of the rules is important.
The operators and their purpose in defining the rules are as
follows:
.TP
.B =   
THIS OPERATOR IS NOT ALLOWED.  If used, and warning message is
printed and it is treated as ==
.TP
.B :=  
Set, this attribute and value will always be placed in the
output A/V Pairs.  If the attribute exists, it is overwritten.
.TP
.B  ==  
Equal, value must match exactly.
.TP
.B  =*  
Always Equal, allow all values for the specified attribute.
.TP
.B    !*  
Never Equal, disallow all values for the specified attribute.
( This is redundant, as any A/V Pair not explicitly permitted
will be dropped ).
.TP
.B    !=  
Not Equal, value must not match.
.TP
.B    >=  
Greater Than or Equal
.TP
.B    <=  
Less Than or Equal
.TP
.B    >   
Greater Than
.TP
.B    <   
Less Than
.PP
If regular expressions are enabled the following operators are
also possible.  ( Regular Expressions are included by default
unless your system doesn't support them, which should be rare ).
The value field uses standard regular expression syntax.
.PP
.TP
.B    =~  
Regular Expression Equal
.TP
.B    !~  
Regular Expression Not Equal
.PP
See the default \fI/etc/raddb/mods-config/attr_filter/\fP for working examples of
sample rule ordering and how to use the different operators.
.DE
.PP
The configuration items are:
.IP file
This specifies the location of the file used to load the filter rules.
This file is used to filter the accounting response, packet before it
is proxied, proxy response from the home server, or our response to
the NAS.
.IP key
Usually %{Realm} (the default).  Can also be %{User-Name}, or other
attribute that exists in the request.  Note that the module always
keys off of attributes in the request, and NOT in any other packet.
.IP relaxed
If set to 'yes', then attributes which do not match any filter rules
explicitly, will also be allowed. This behaviour may be overridden 
for an individual filter block using the Relax-Filter check item.
The default for this configuration item is 'no'.
.PP
.SH SECTIONS
.IP preacct
Filters Accounting-Request packets.
.IP accounting
Filters Accounting-Response packets.
.IP pre-proxy
Filters Accounting-Request or Access-Request packets prior to proxying
them.
.IP post-proxy
Filters Accounting-Response, Access-Accept, Access-Reject, or
Access-Challenge responses from a home server.
.IP authorize
Filters Access-Request packets.
.IP post-auth
Filters Access-Accept or Access-Reject packets.
.PP
.SH FILES
.I /etc/raddb/radiusd.conf
.I /etc/raddb/mods-config/attr_filter/*
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5)
.SH AUTHOR
Chris Parker, cparker@segv.org



doc/ChangeLog
FreeRADIUS 3.0.15 Fri 26 May 2017 13:00:00 EDT urgency=medium
	Feature improvements
	* Provide HOSTNAME in default systemd files.
	* Incorporate RedHat specific files
	* Update dictionary.starent, dictionary.ruckus
	* Allow builds without TCP or DHCP

	Bug fixes
	* don't crash when reading corrupted data from session
	  resumption cache.  Fixes #1999.
	* Parse port in dhcpclient.  Fixes #2000.
	* Don't leak memory for OpenSSL.
	  Patch from Guido Vranken.
	* Portability fixes taken from OpenBSD port collection.
	* run rad_authlog after post-auth for Access-Reject.
	* Don't process VMPS packets twice.

FreeRADIUS 3.0.14 Fri 26 May 2017 13:00:00 EDT urgency=medium
	Feature improvements
	* Enforce TLS client certificate expiration on
	  session resumption, and Session-Timeout.
	  See CVE-2017-9148.
	* Updated dictionary.cisco.vpn3000, dictionary.patton
	* Added dictionary.dellemc
	* Lowered the log output for failed PEAP sessions.
	* ALlow utc in rlm_date.  Patch from
	  Peter Lambrechtsen.
	* The internal OpenSSL session cache has been
	  disabled.  Please see mods-available/eap
	* Update detail reader documentation.
	  Patch from Matthew Newton.  Fixes #1973.
	* Make outgoing RadSec connections non-blocking.
	* Add SQL backing to Moonshot-*-TargetedId
	  generation.  Patch from Stefan Paetow.

	Bug fixes
	* radtest uses Cleartext-Password for EAP, not
	  User-Password.
	* Update documentation for mods-enabled/ linking.
	* Enhanced checks for moonshot salt.  Fixes #1933.
	* Allow session resumption for RadSec connections.
	  Fixes #1936.
	* Update "huntgroups" file to note that port ranges
	  are not supported.
	* Fix OpenSSL permissions issues on default key files.
	  Fixes #1941.
	* Certificates are not required when PSK is used.
	* Allow SubjectAltName as first extension in cert.
	  Fixes #1946.
	* Fixed talloc issue with TLS session resumption.
	  Fixes #1980.
	* "&Attr-26 := 0x01" now produces useful error messages.
	* Handle connection error in rlm_ldap_cacheable_groupobj.
	  Fixes #1951.
	* Fix endian issues in DHCP.
	* Multiple minor fixes for Coverity complaints.
	* Handle unexpected regex.  Fixes #1959.
	* Fix minor issues in dictionaries.
	* Fix typos and grammar.  Patches from Alan Buxey.
	* Fix erroneous VP creation in rlm_preproces.
	* Fix MIB.  Patch from Jeff Gehlbach.
	* Trust router updates from Alejandro Perez.
	* Allow build with LibreSSL.  Fixes #1989
	* Use correct packet for channel bindings.  Fixes #1990.
	* Many fixes found by PVS-Studio.  Thanks to PVS-Studio
	  for giving us a test license.  Please see the git commit
	  history for more information.
	* Fix incorrect length check in EAP-PWD.  This may
	  be exploitable.

FreeRADIUS 3.0.13 Mon 06 Mar 2017 13:00:00 EDT urgency=medium
	Feature improvements
	* Add dictionary.rfc7930.  Note that we do not implement
	  the RFC.
	* Added 'cipher_server_preference' to mods-available/eap
	  Patch from #1797.
	* OpenSSL 1.1.0 compatibility fixes.
	* rlm_perl: radiusd::xlat to evaluate xlat string
	  within perl script
	* Allow authentication retry in winbind. Patch from
	  Herwin Weststrate. See raddb/mods-available/mschap.
	* Added "recv-coa" method to rlm_rest.  It behaves the
	  same as "authorize".
	* Document Trust Router tr_port option.  Patch from
	  Stefan Paetow.
	* Update elasticsearch/logstash examples so that they work
	  with elastic stack v5.  Patch from Matthew Newton.
	* Print information about packets, replies, and contents
	  in the detail file reader.
	* Update abfab-tr policy.  Pull request #1893
	  from Stefan Paetow.
	* Reject packets which contain User-Password and
	  EAP-Message.
	* Add example for filtering Access-Challenge.
	  See sites-enabled/default.
	* Pull symlink fixes from v4.0.x.  Fixes #1859.
	* Add systemd reload.  Not everything is reloaded, but
	  some is.  Fixes #1662.
	* Better documentation for listen "ipaddr".  Fixes #1921
	* Add dictionary.cnergee, updated dictionary.nomadix.
	* radclient no longer needs -x to print statistics with -s.

	Bug fixes
	* Minor typos.  Fixes #1763
	* Fix typo in RPM build.  Closes #1767.
	* rlm_mschap check for password expiry only
	  if password was correct.  Fixes #1762.
	* Update debian build.
	* update rlm_counter "man" page.  Fixes #1775.
	* Remove erroneous assert.  Fixes #1778.
	* fix mschap password change test.  Fixes #1792.
	* Cleanup config file on data remove.  Fixes #1795.
	* passwd module returns "notfound" if not found.
	* Check for old OpenSSL, and don't build rlm_eap_fast
	  if it necessary.  Fixes #1803
	* Cleanup memory better after ldap version query.
	  Patch from Aleksey Katargin.
	* Rename lt_* functions to avoid linker issues with
	  libtool.  Fixes #1277
	* Many miscellaneous fixes and typos.
	* Allow long strings in %{%{foo} bar:-%{baz} blah".
	  Fixes #1866
	* Fix filtering operators, along with more documentation and
	  more tests for them.
	* Fix OpenSSL fixes.  Fixes #1876.
	* Finish SQL select queries even when SELECT returns no rows.
	  Fixes #1879.
	* Set Module-Failure-Message for more EAP errors.
	* Correct typo in dictionary.rfc5580.  Fixes #1882
	* Remove obselete systemd syslog.target.
	* Client-Port-Balance load-balancing now uses client port.
	* Radrelay examples fixed from Alex Clouter.
	* Update systemd target.  Pull request #1896.
	* Trim starting whitespace in xlat strings.
	* Get MySQL result lengths using normal API.
	* suid down after fchown().  Fixes #1914.
	* Fix cases of comparing pointer to NUL character.  Fixes #1915.
	* OpenSSL v1.1 fixes.  Pull request #1921.
	* Better Handle v4/v6 host names.  Pull request #1919.
	* Remove "Auth-Type = System" from docs and examples.
	* Don't crash on malformed %{home_server}.  Fixes #1922
	* fix erroneous use of talloc destructor in rlm_eap
	* Issue trigger modules.sql.fail.  Fixes #1923
	* Document python_path gotcha's.  Fixes #1845
	* dlopen() the specific version of Python.  Fixes #1592

FreeRADIUS 3.0.12 Thur 29 Sep 2016 13:00:00 EDT urgency=medium

src/modules/rlm_attr_filter/rlm_attr_filter.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_attr_filter.c
 * @brief Filter the contents of a list, allowing only certain attributes.
 *
 * @copyright (C) 2001,2006 The FreeRADIUS server project
 * @copyright (C) 2001 Chris Parker <cparker@starnetusa.net>
 */
RCSID("$Id$")

#include	<freeradius-devel/radiusd.h>
#include	<freeradius-devel/modules.h>
#include	<freeradius-devel/rad_assert.h>

#include	<sys/stat.h>

#include	<ctype.h>
#include	<fcntl.h>

/*
 *	Define a structure with the module configuration, so it can
 *	be used as the instance handle.
 */
typedef struct rlm_attr_filter {
	char const	*filename;
	vp_tmpl_t	*key;
	bool		relaxed;
	PAIR_LIST	*attrs;
} rlm_attr_filter_t;

static const CONF_PARSER module_config[] = {
	{ FR_CONF_OFFSET("filename", FR_TYPE_FILE_INPUT | FR_TYPE_REQUIRED, rlm_attr_filter_t, filename) },
	{ FR_CONF_OFFSET("key", FR_TYPE_TMPL, rlm_attr_filter_t, key), .dflt = "&Realm", .quote = T_BARE_WORD },
	{ FR_CONF_OFFSET("relaxed", FR_TYPE_BOOL, rlm_attr_filter_t, relaxed), .dflt = "no" },
	CONF_PARSER_TERMINATOR
};

static void check_pair(REQUEST *request, VALUE_PAIR *check_item, VALUE_PAIR *reply_item, int *pass, int *fail)
{
	int compare;

	if (check_item->op == T_OP_SET) return;

	compare = fr_pair_cmp(check_item, reply_item);
	if (compare < 0) {
		RPEDEBUG("Comparison failed");
	}

	if (compare == 1) {
		++*(pass);
	} else {
		++*(fail);
	}

	if (RDEBUG_ENABLED3) {
		char rule[1024], pair[1024];

		fr_pair_snprint(rule, sizeof(rule), check_item);
		fr_pair_snprint(pair, sizeof(pair), reply_item);
		RDEBUG3("%s %s %s", pair, compare == 1 ? "allowed by" : "disallowed by", rule);
	}

	return;
}

static int attr_filter_getfile(TALLOC_CTX *ctx, char const *filename, PAIR_LIST **pair_list)
{
	vp_cursor_t cursor;
	int rcode;
	PAIR_LIST *attrs = NULL;
	PAIR_LIST *entry;
	VALUE_PAIR *vp;

	rcode = pairlist_read(ctx, filename, &attrs, 1);
	if (rcode < 0) {
		return -1;
	}

	/*
	 * Walk through the 'attrs' file list.
	 */

	entry = attrs;
	while (entry) {
		entry->check = entry->reply;
		entry->reply = NULL;

		for (vp = fr_pair_cursor_init(&cursor, &entry->check);
		     vp;
		     vp = fr_pair_cursor_next(&cursor)) {
		    /*
		     * If it's NOT a vendor attribute,
		     * and it's NOT a wire protocol
		     * and we ignore Fall-Through,
		     * then bitch about it, giving a good warning message.
		     */
		     if ((vp->da->vendor == 0) &&
			 (vp->da->attr > 1000)) {
			WARN("[%s]:%d Check item \"%s\"\n\tfound in filter list for realm \"%s\".\n",
			       filename, entry->lineno, vp->da->name, entry->name);
		    }
		}

		entry = entry->next;
	}

	*pair_list = attrs;
	return 0;
}


/*
 *	(Re-)read the "attrs" file into memory.
 */
static int mod_instantiate(void *instance, UNUSED CONF_SECTION *conf)
{
	rlm_attr_filter_t *inst = instance;
	int rcode;

	rcode = attr_filter_getfile(inst, inst->filename, &inst->attrs);
	if (rcode != 0) {
		ERROR("Errors reading %s", inst->filename);

		return -1;
	}

	return 0;
}


/*
 *	Common attr_filter checks
 */
static rlm_rcode_t CC_HINT(nonnull(1,2)) attr_filter_common(void const *instance, REQUEST *request,
							    RADIUS_PACKET *packet)
{
	rlm_attr_filter_t const *inst = instance;
	VALUE_PAIR	*vp;
	vp_cursor_t	input, check, out;
	VALUE_PAIR	*input_item, *check_item, *output;
	PAIR_LIST	*pl;
	int		found = 0;
	int		pass, fail = 0;
	char const	*keyname = NULL;
	char		buffer[256];
	ssize_t		slen;

	if (!packet) return RLM_MODULE_NOOP;

	slen = tmpl_expand(&keyname, buffer, sizeof(buffer), request, inst->key, NULL, NULL);
	if (slen < 0) return RLM_MODULE_FAIL;
	if ((keyname == buffer) && is_truncated((size_t)slen, sizeof(buffer))) {
		REDEBUG("Key too long, expected < " STRINGIFY(sizeof(buffer)) " bytes, got %zi bytes", slen);
		return RLM_MODULE_FAIL;
	}

	/*
	 *	Head of the output list
	 */
	output = NULL;
	fr_pair_cursor_init(&out, &output);

	/*
	 *      Find the attr_filter profile entry for the entry.
	 */
	for (pl = inst->attrs; pl; pl = pl->next) {
		int fall_through = 0;
		int relax_filter = inst->relaxed;

		/*
		 *  If the current entry is NOT a default,
		 *  AND the realm does NOT match the current entry,
		 *  then skip to the next entry.
		 */
		if ((strcmp(pl->name, "DEFAULT") != 0) &&
		    (strcmp(keyname, pl->name) != 0))  {
		    continue;
		}

		RDEBUG2("Matched entry %s at line %d", pl->name, pl->lineno);
		found = 1;

		for (check_item = fr_pair_cursor_init(&check, &pl->check);
		     check_item;
		     check_item = fr_pair_cursor_next(&check)) {
			if (!check_item->da->vendor &&
			    (check_item->da->attr == FR_FALL_THROUGH) &&
				(check_item->vp_uint32 == 1)) {
				fall_through = 1;
				continue;
			}
			else if (!check_item->da->vendor && check_item->da->attr == FR_RELAX_FILTER) {
				relax_filter = check_item->vp_uint32;
				continue;
			}

			/*
			 *    If it is a SET operator, add the attribute to
			 *    the output list without checking it.
			 */
			if (check_item->op == T_OP_SET ) {
				vp = fr_pair_copy(packet, check_item);
				if (!vp) {
					goto error;
				}
				xlat_eval_do(request, vp);
				fr_pair_cursor_append(&out, vp);
			}
		}

		/*
		 *	Iterate through the input items, comparing
		 *	each item to every rule, then moving it to the
		 *	output list only if it matches all rules
		 *	for that attribute.  IE, Idle-Timeout is moved
		 *	only if it matches all rules that describe an
		 *	Idle-Timeout.
		 */
		for (input_item = fr_pair_cursor_init(&input, &packet->vps);
		     input_item;
		     input_item = fr_pair_cursor_next(&input)) {
			pass = fail = 0; /* reset the pass,fail vars for each reply item */

			/*
			 *  Reset the check_item pointer to beginning of the list
			 */
			for (check_item = fr_pair_cursor_first(&check);
			     check_item;
			     check_item = fr_pair_cursor_next(&check)) {
				/*
				 *  Vendor-Specific is special, and matches any VSA if the
				 *  comparison is always true.
				 */
				if ((check_item->da->attr == FR_VENDOR_SPECIFIC) && (input_item->da->vendor != 0) &&
				    (check_item->op == T_OP_CMP_TRUE)) {
					pass++;
					continue;
				}

				if (input_item->da == check_item->da) {
					check_pair(request, check_item, input_item, &pass, &fail);
				}
			}

			RDEBUG3("Attribute \"%s\" allowed by %i rules, disallowed by %i rules",
				input_item->da->name, pass, fail);
			/*
			 *  Only move attribute if it passed all rules, or if the config says we
			 *  should copy unmatched attributes ('relaxed' mode).
			 */
			if (fail == 0 && (pass > 0 || relax_filter)) {
				if (!pass) {
					RDEBUG3("Attribute \"%s\" allowed by relaxed mode", input_item->da->name);
				}
				vp = fr_pair_copy(packet, input_item);
				if (!vp) {
					goto error;
				}
				fr_pair_cursor_append(&out, vp);
			}
		}

		/* If we shouldn't fall through, break */
		if (!fall_through) {
			break;
		}
	}

	/*
	 *	No entry matched.  We didn't do anything.
	 */
	if (!found) {
		rad_assert(!output);
		return RLM_MODULE_NOOP;
	}

	/*
	 *	Replace the existing request list with our filtered one
	 */
	fr_pair_list_free(&packet->vps);
	packet->vps = output;

	if (request->packet->code == FR_CODE_ACCESS_REQUEST) {
		request->username = fr_pair_find_by_num(request->packet->vps, 0, FR_STRIPPED_USER_NAME, TAG_ANY);
		if (!request->username) {
			request->username = fr_pair_find_by_num(request->packet->vps, 0, FR_USER_NAME, TAG_ANY);
		}
		request->password = fr_pair_find_by_num(request->packet->vps, 0, FR_USER_PASSWORD, TAG_ANY);
	}

	return RLM_MODULE_UPDATED;

	error:
	fr_pair_list_free(&output);
	return RLM_MODULE_FAIL;
}

#define RLM_AF_FUNC(_x, _y) static rlm_rcode_t CC_HINT(nonnull) mod_##_x(void *instance, UNUSED void *thread, REQUEST *request) \
	{ \
		return attr_filter_common(instance, request, request->_y); \
	}

RLM_AF_FUNC(authorize, packet)
RLM_AF_FUNC(post_auth, reply)

RLM_AF_FUNC(preacct, packet)
RLM_AF_FUNC(accounting, reply)

#ifdef WITH_PROXY
RLM_AF_FUNC(pre_proxy, proxy->packet)
RLM_AF_FUNC(post_proxy, proxy->reply)
#endif

#ifdef WITH_COA
RLM_AF_FUNC(recv_coa, packet)
RLM_AF_FUNC(send_coa, reply)
#endif

/* globally exported name */
extern rad_module_t rlm_attr_filter;
rad_module_t rlm_attr_filter = {
	.magic		= RLM_MODULE_INIT,
	.name		= "attr_filter",
	.inst_size	= sizeof(rlm_attr_filter_t),
	.config		= module_config,
	.instantiate	= mod_instantiate,
	.methods = {
		[MOD_AUTHORIZE]		= mod_authorize,
		[MOD_PREACCT]		= mod_preacct,
		[MOD_ACCOUNTING]	= mod_accounting,
#ifdef WITH_PROXY
		[MOD_PRE_PROXY]		= mod_pre_proxy,
		[MOD_POST_PROXY]	= mod_post_proxy,
#endif
		[MOD_POST_AUTH]		= mod_post_auth,
#ifdef WITH_COA
		[MOD_RECV_COA]		= mod_recv_coa,
		[MOD_SEND_COA]		= mod_send_coa
#endif
	},
};

